### DISCUSSION TITLE
Multiple bugs

### DISCUSSION THREAD

**Discussion by Bananeweizen** [Genuitec/Copilot4Eclipse-community#53] [Help] [answered]
* com.genuitec.copilot4eclipse 1.3.2.202411110916 (most recent version) requires a very narrow version range of `org.eclipse.wild
 webdeveloper.embedder.node;bundle-version="[1.0.2,1.0.4)"` in its manifest. That node embedder plugin is one year old, see the dates in the version numbers: 
![image](https://github.com/user-attachments/assets/b87bc43f-f652-473b-8a85-3307b5ab3d6b)
 Current versions of Eclipse contain a much more recent wildwebdeveloper feature. That's why the current version of your plugin fails to install into Eclipse IDEs that already contain another plugin depending on a more recent node embedder plugin:
 ```
ERROR: org.eclipse.equinox.p2.director code=10053 Cannot complete the install because one or more required items could not be found.
  at org.eclipse.oomph.util.OomphPlugin.coreException(OomphPlugin.java:296)
  at org.eclipse.oomph.p2.internal.core.ProfileTransactionImpl.resolve(ProfileTransactionImpl.java:446)
  at org.eclipse.oomph.p2.internal.core.ProfileTransactionImpl.commit(ProfileTransactionImpl.java:356)
  at org.eclipse.oomph.setup.p2.impl.P2TaskImpl.perform(P2TaskImpl.java:918)
  at org.eclipse.oomph.setup.internal.core.SetupTaskPerformer.doPerformNeededSetupTasks(SetupTaskPerformer.java:3869)
  at org.eclipse.oomph.setup.internal.core.SetupTaskPerformer.performNeededSetupTasks(SetupTaskPerformer.java:3797)
  at org.eclipse.oomph.setup.internal.core.SetupTaskPerformer.performTriggeredSetupTasks(SetupTaskPerformer.java:3778)
  at org.eclipse.oomph.setup.internal.core.SetupTaskPerformer.perform(SetupTaskPerformer.java:3656)
  at org.eclipse.oomph.setup.ui.wizards.ProgressPage$9.run(ProgressPage.java:592)
  at org.eclipse.oomph.setup.ui.wizards.ProgressPage$11$1.run(ProgressPage.java:721)
  at org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)
  ERROR: org.eclipse.equinox.p2.director code=0 Software being installed: artificial_root 1.0.0.v1731934985510
  ERROR: org.eclipse.equinox.p2.director code=0 Missing requirement: Copilot4Eclipse 1.3.2.202411110916 (com.genuitec.copilot4eclipse 1.3.2.202411110916) requires 'osgi.bundle; org.eclipse.wildwebdeveloper.embedder.node [1.0.2,1.0.4)' but it could not be found
  ERROR: org.eclipse.equinox.p2.director code=1 Cannot satisfy dependency:
    ERROR: org.eclipse.equinox.p2.director code=0 From: artificial_root 1.0.0.v1731934985510
    ERROR: org.eclipse.equinox.p2.director code=0 To: org.eclipse.equinox.p2.iu; com.genuitec.copilot4eclipse.feature.feature.group [1.2.2,9.9.0)
  ERROR: org.eclipse.equinox.p2.director code=1 Cannot satisfy dependency:
    ERROR: org.eclipse.equinox.p2.director code=0 From: Copilot4Eclipse Core 1.3.2.202411110916 (com.genuitec.copilot4eclipse.core.feature.feature.group 1.3.2.202411110916)
    ERROR: org.eclipse.equinox.p2.director code=0 To: org.eclipse.equinox.p2.iu; com.genuitec.copilot4eclipse [1.3.2.202411110916,1.3.2.202411110916]
  ERROR: org.eclipse.equinox.p2.director code=1 Cannot satisfy dependency:
    ERROR: org.eclipse.equinox.p2.director code=0 From: Copilot4Eclipse 1.3.2.202411110916 (com.genuitec.copilot4eclipse.feature.feature.group 1.3.2.202411110916)
    ERROR: org.eclipse.equinox.p2.director code=0 To: org.eclipse.equinox.p2.iu; com.genuitec.copilot4eclipse.core.feature.feature.group
```
* com.genuitec.copilot4eclipse exports the `com.genuitec.copilot4eclipse.utils` package. It should not do that, as that package is already exported by the plugin com.genuitec.copilot4eclipse.utils. The 2 exported packages are considered as 2 _different_ packages by Eclipse.
![image](https://github.com/user-attachments/assets/8f4b6681-b648-45e0-b1e2-303f52396dcb)
* Please stop changing the URLs of view icons in plugin updates. That breaks the icon in every workspace which has one of the views opened. Eclipse persists the image URL of open views in the workspace metadata and tries to load those images again without activating the related bundles and without checking for updates in the image URL. With your previous changes of icons everyone got red "image not found" icons in all the Copilot related views until they were actively opened again.

**Comment by wayneparrott** [comment]
@Bananeweizen 

Let me start with I generally appreciate constructive feedback. Most of your feedback is well taken and know that we are working to continually improve Copilot4Eclipse. Maybe it's me having a bad rainy Monday or maybe your lecture about node-embedder versions and constraints is annoying me atm. I guess the 3000+ successful installs of Copilot4Eclipse 1.3 this month (50,000+ total) in multiple Eclipse distros back to 2023-06 has something to do with it.

Moving forward, I would like to replicate the issue you are reporting. Please share the details of your Eclipse environment, JRE, ... 3rd party plugins, etc so I can setup an exact version of your environment. Also share a screenshot of the update-sites you have enabled/disabled as this is probably where something is preventing the installation. With that info, I'll see if I can replicate the issue. If there is something breaking on the Copilot4Eclipse side we will address it asap.

Regards,
Wayne
Copilot4Eclipse Team

**Reply by Bananeweizen** [reply]
Sorry Wayne, it was not my intention to annoy you or to lecture something. I was just in a hurry to get this posted with as much information as I had, while other tasks were already waiting.

Meanwhile I assume the first bullet point is plain wrong. I now checked some more things:  the node embedder plugins are not singletons, therefore there should not be any conflict between multiple versions. That also fits to me having multiple versions of that plugin in a recent committers IDE with the copilot plugin installed. And last but not least: The error log should also mention the _conflicting_ dependency if there were a conflict. Therefore I now believe the above error was just and only a network problem. As I'm not sure whether the copilot update site contains a copy of the embedder feature or just a reference to the eclipse.org update site, I probably can't do any further checks how this error came up. Sorry for wasting your time there.

**Reply by wayneparrott** [reply, marked as answer]
Thx for following up. I hope Copilot4Eclipse serves you well. We are currently improving the plugin architecture and developing a new level of UI capability not seen in Eclipse before. As part of this effort we are looking to tighten up the tolerances across the product and will be more diligent in our methods thanks to some of your suggestions.
